pkg/controllers/mcs/mcs_serviceexport_controller.go

```diff
@@ -0,0 +1,200 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	mcsv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	networkingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/networking/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/events"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+const (
+	// MultiClusterServiceControllerName MultiClusterServiceController name.
+	MultiClusterServiceControllerName = "multi-cluster-service-controller"
+
+	// MultiClusterServiceControllerFinalizer is added to MultiClusterService to ensure related Works are deleted
+	// before MultiClusterService itself is deleted.
+	MultiClusterServiceControllerFinalizer = "karmada.io/multi-cluster-service-controller"
+)
+
+// MultiClusterServiceController lists/watches the multiClusterService resource,
+// generates ServiceExport and propagate it to the source clusters,
+// and generates ServiceImport and propagate it to the destination clusters.
+// Note: This controller only focus on  multiClusterService of type CrossCluster.
+type MultiClusterServiceController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+func (c *MultiClusterServiceController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling MultiClusterService %s.", req.NamespacedName.String())
+
+	mcs := &networkingv1alpha1.MultiClusterService{}
+	if err := c.Client.Get(ctx, req.NamespacedName, mcs); err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	ownerSvc := &corev1.Service{}
+	err := c.Client.Get(ctx, types.NamespacedName{Name: mcs.Name, Namespace: mcs.Namespace}, ownerSvc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if err := c.reconcileServiceExport(ctx, mcs, ownerSvc); err != nil {
+		klog.Errorf("Failed to build ServiceExport for MultiClusterService(%s), error: %v", req.NamespacedName.String(), err)
+		return controllerruntime.Result{}, err
+	}
+	if !mcs.DeletionTimestamp.IsZero() || !mcs.Spec.HasExposureTypeCrossCluster() {
+		return c.removeFinalizer(mcs)
+	}
+	// ensure related Works are deleted before MultiClusterService itself is deleted.
+	return c.ensureFinalizer(mcs)
+}
+func (c *MultiClusterServiceController) deriveServiceExportFromMultiClusterService(
+	ctx context.Context, mcs *networkingv1alpha1.MultiClusterService, ownSvc *corev1.Service) error {
+	newSvcExport := &mcsv1alpha1.ServiceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: mcs.Namespace,
+			Name:      mcs.Name,
+			Labels: map[string]string{
+				networkingv1alpha1.MultiClusterServiceDerivedCrossClusterKind: "",
```

Comment:
Do we really need this label? We can add the finalizer.

Reply:
The current core idea of PR is to use the feature of dependency distribution to set serviceexport as a dependent resource of svc. If we add a finalizer to serviceexport, it would be difficult to determine when this finalizer should be removed.